### PR TITLE
Spark 3.5: Parallelize reading files in snapshot and migrate procedures

### PR DESCRIPTION
**Pull request by manuzhang**
Similar to #9274 for `snapshot` and `migrate` procedures.

### CODE REVIEW

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSnapshotTableProcedure.java [hunk 1]
```diff
@@ -223,4 +223,31 @@ public void testInvalidSnapshotsCases() throws IOException {
...
+    Table createdTable = validationCatalog.loadTable(tableIdent);
+    String tableLocation = createdTable.location();
+    assertThat(tableLocation)
+        .as("Table should not have the original location")
+        .isNotEqualTo(location);
+
+    assertEquals(
+        "Should have expected rows",
+        ImmutableList.of(row(1L, "a"), row(2L, "b")),
+        sql("SELECT * FROM %s ORDER BY id", tableName));
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSnapshotTableProcedure.java:251]:
```suggestion
        assertThat(sql("SELECT * FROM %s ORDER BY id", tableName))).containsExactly(row(1L, "a"), row(2L, "b"));
```

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/SnapshotTableSparkAction.java [hunk 2]
```diff
@@ -98,6 +99,12 @@ public SnapshotTableSparkAction tableProperty(String property, String value) {
     return this;
   }
 
+  @Override
+  public SnapshotTable parallelism(int num) {
```
Comment (nastra) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/SnapshotTableSparkAction.java:103]:
```suggestion
  public SnapshotTable parallelism(int numThreads) {
```

File: api/src/main/java/org/apache/iceberg/actions/MigrateTable.java [hunk 3]
```diff
@@ -60,6 +60,16 @@ default MigrateTable backupTableName(String tableName) {
     throw new UnsupportedOperationException("Backup table name cannot be specified");
   }
 
+  /**
+   * Sets the number of threads to use for file reading. The default is 1.
+   *
+   * @param num the number of threads
+   * @return this for method chaining
+   */
+  default MigrateTable parallelism(int num) {
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:69]:
```suggestion
  default MigrateTable parallelism(int numThreads) {
```

File: api/src/main/java/org/apache/iceberg/actions/SnapshotTable.java [hunk 4]
```diff
@@ -57,6 +57,16 @@ public interface SnapshotTable extends Action<SnapshotTable, SnapshotTable.Resul
    */
   SnapshotTable tableProperty(String key, String value);
 
+  /**
+   * Sets the number of threads to use for file reading. The default is 1.
+   *
+   * @param num the number of threads
+   * @return this for method chaining
+   */
+  default SnapshotTable parallelism(int num) {
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/actions/SnapshotTable.java:66]:
```suggestion
  default SnapshotTable parallelism(int numThreads) {
```

File: docs/docs/spark-procedures.md [hunk 5]
```diff
@@ -588,6 +589,8 @@ By default, the original table is retained with the name `table_BACKUP_`.
 | `properties`  | ️   | map<string, string> | Properties for the new Iceberg table |
 | `drop_backup` |   | boolean | When true, the original table will not be retained as backup (defaults to false) |
 | `backup_table_name` |  | string | Name of the table that will be retained as backup (defaults to `table_BACKUP_`) |
+| `parallelism` |   | int | Number of threads to use for file reading (defaults to 1) |
+
```
Comment (nastra) [inline docs/docs/spark-procedures.md:593]: nit: unnecessary newline

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java [hunk 6]
```diff
@@ -232,4 +232,20 @@ public void testMigrateEmptyTable() throws Exception {
     Object result = scalarSql("CALL %s.system.migrate('%s')", catalogName, tableName);
     assertThat(result).isEqualTo(0L);
   }
+
+  @TestTemplate
+  public void testMigrateWithParallelism() throws IOException {
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:237]: can you please also add some tests where `parallelism` is -1 / 0 / <some large value>

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java [hunk 7]
```diff
@@ -232,4 +232,25 @@ public void testMigrateEmptyTable() throws Exception {
...
+    assumeThat(catalogName).isEqualToIgnoringCase("spark_catalog");
+    String location = Files.createTempDirectory(temp, "junit").toFile().toString();
+    for (int p = -1; p <= 2; p++) {
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: what I meant is to have a separate test method where the parallelism is -1 / 0 / 10 or so

Comment (nastra) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: I haven't looked at the impl, but what happens if parallelism is set to 10?

Comment (manuzhang) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: 10 threads are used for executor service

Comment (nastra) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: in that case let's remove the loop here and use numThreads = 5 or so

Comment (manuzhang) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: Don't we want to test passed-in parallelism of `-1` and `0`? We don't forbid it so I think it's valid case.

Comment (nastra) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: yes but we don't want to do it in a loop. If something fails, it won't be clear with which parallelism parameter it failed. Using loops inside tests is a sign that this should be either a parameterized test or just a separate test method with a different input

Comment (manuzhang) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: I was trying that way initially, but there's no way to add parameters for just one test.

Comment (RussellSpitzer) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: Why should 0 and -1 pass? (Also I agree the for loop is a bit confusing here, also try not to use single letter variables)

Comment (RussellSpitzer) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: Quick test style as well, if I want to test multiple cases and the framework doesn't support parameterization of the individual tests I would do something like

```
def body(parameter) {
   outcome =  thing()
   assert expected = outcome 
}
```

```
def testIllegal() {
   assert(body(-1)).throws
}


def testValid() {
   body(1)
   body(10)
}
```

Comment (manuzhang) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:240]: > Why should 0 and -1 pass?

@RussellSpitzer because we only use executor service when passed-in parallelism is larger than 1 and use main thread otherwise. 

Let me explore your suggested approach.

Comment (nastra) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java]: > I was trying that initially, but there's no way to add parameters for just one test.

This is because the test is already parameterized at the class level

Comment (manuzhang) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java]: Thanks @nastra , I've updated as @RussellSpitzer suggests.

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java [hunk 8]
```diff
@@ -171,6 +176,10 @@ public static List<DataFile> listPartition(
     }
   }
 
+  public static boolean isFileReadingParallelized() {
```
Comment (nastra) [inline data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:179]: I don't think it's a good idea to introduce this, since this reflects whatever was set after calling `listPartition(...)` and depending on the order of calls, it could return something unexpected

Comment (manuzhang) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:179]: @nastra do you have an idea to test whether parallelism setting is working?

Comment (nastra) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:179]: I haven't checked the details on how to properly test this, but introducing a static variable isn't the right way

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/ProcedureUtil.java [hunk 9]
```diff
@@ -51,4 +56,29 @@ static String statsFileLocation(String tableLocation) {
...
+          new ThreadFactoryBuilder().build());
+    }
+
+    @Override
+    public void execute(@NotNull Runnable task) {
+      super.execute(task);
+      executedTasks++;
+    }
+
+    public int getExecutedTasks() {
```
Comment (nastra) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/ProcedureUtil.java:80]:
```suggestion
    public int executedTasks() {
```

**Review by RussellSpitzer — COMMENTED** [review]
The code here IMHO should be checking what number of threads is passed to the executor service constructor. That's sufficient to test for me. 

I also think that we shouldn't allow 0 or -1. Seems like those should have been forbidden before and I don't see why we would allow them now.

I'm also approving this in terms that when Eduard is satisfied i'm good to go as well.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java [hunk 10]
```diff
@@ -232,4 +237,42 @@ public void testMigrateEmptyTable() throws Exception {
...
+  private void testWithParallelism(int parallelism) throws IOException {
+    String location = Files.createTempDirectory(temp, "junit").toFile().toString();
+    sql(
+        "CREATE TABLE %s (id bigint NOT NULL, data string) USING parquet LOCATION '%s'",
+        tableName, location);
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
+    sql("INSERT INTO TABLE %s VALUES (2, 'b')", tableName);
+
+    try (MockedStatic<MoreExecutors> executors = mockStatic(MoreExecutors.class)) {
+      ProcedureUtil.TestExecutorService testService = new ProcedureUtil.TestExecutorService();
```
[the remark this answers is not in the document]
Comment (manuzhang) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMigrateTableProcedure.java:260]: `TestExecutorService` is to verify an executor service is used when `parallelism` for `migrate` procedure is larger than 1. The number of threads used by `TestExecutorService` doesn't matter here.

**Review by aokolnychyi — APPROVED** [review]
LGTM. Nothing to add except what was already noted by others.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/BaseProcedure.java [hunk 11]
```diff
@@ -237,4 +237,11 @@ protected ExecutorService executorService(int threadPoolSize, String nameFormat)
 
     return executorService;
   }
+
+  protected int checkAndGetArg(
```
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/BaseProcedure.java:241]: If we want to add this logic for validating the parser, we should do this via `ProcedureInput`, not as a separate method. My suggestion would be to revert this change and use the original approach in this PR. In a separate PR, we can switch MIGRATE and SNAPSHOT procedures to use `ProcedureInput` and add the validation there.

File: data/src/main/java/org/apache/iceberg/data/MigrationService.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+ */
+package org.apache.iceberg.data;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadPoolExecutor;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.MoreExecutors;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/** Have a separate class for getting ExecutorService to make it testable with static mock */
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/MigrationService.java:27]: Are we sure it is a good idea to add another public class for this? How do we see it being used in the future? Why not use `ThreadPools` instead?

Comment (manuzhang) [reply data/src/main/java/org/apache/iceberg/data/MigrationService.java:27]: When we do static mock, all invocations need to be mocked as well. That includes `ThreadPools` used for `SnapshotProducer` and other places. Having a separate class and mock it is the cleanest way.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/BaseProcedure.java (hunk 11 above)
Comment (aokolnychyi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/BaseProcedure.java:241]: Is this method being used or shall we remove it?

File: data/src/main/java/org/apache/iceberg/data/MigrationService.java (hunk 12 above)
Comment (aokolnychyi) [reply data/src/main/java/org/apache/iceberg/data/MigrationService.java:27]: I am not convinced it justifies adding another public class. What if we overload `listPartition` or whatever method we need in `TableMigrationUtil` to accept `ExecutorService`?

It is an open question whether we should accept `ExecutorService` or `int` in the actions (not procedures), though.

Comment (manuzhang) [reply data/src/main/java/org/apache/iceberg/data/MigrationService.java]: @aokolnychyi I finally get time to refactor a bit, accepting `ExecutorService` in the actions, `SparkTableUtil` and  `TableMigrationUtil`.

**Comment by manuzhang** [comment]
@nastra @RussellSpitzer @aokolnychyi Could you please take another look?

**Comment by nastra** [comment]
@manuzhang could you rebase the PR please as there were some changes that might affect it

**Comment by manuzhang** [comment]
@nastra Done!

File: api/src/main/java/org/apache/iceberg/actions/SnapshotTable.java [hunk 13]
```diff
@@ -57,6 +58,17 @@ public interface SnapshotTable extends Action<SnapshotTable, SnapshotTable.Resul
...
   SnapshotTable tableProperty(String key, String value);
 
+  /**
+   * Sets the executor service to use for parallel file reading. The default is not using executor
+   * service.
+   *
+   * @param service executor service
+   * @return this for method chaining
+   */
+  default SnapshotTable executeSnapshotWith(ExecutorService service) {
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/actions/SnapshotTable.java:68]: why not just name it `executeWith` here and in the other API?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java [hunk 14]
```diff
@@ -550,7 +650,7 @@ private static void importUnpartitionedSparkTable(
       }
 
       AppendFiles append = targetTable.newAppend();
-      files.forEach(append::appendFile);
+      files.stream().filter(java.util.Objects::nonNull).forEach(append::appendFile);
```
Comment (nastra) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:653]: can you elaborate why this change is necessary?

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java [hunk 15]
```diff
@@ -215,11 +250,7 @@ private static DataFile buildDataFile(
         .build();
   }
 
-  private static ExecutorService migrationService(int parallelism) {
-    return MoreExecutors.getExitingExecutorService(
```
Comment (nastra) [inline data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:219]: why are we switching here from an exiting executor service?

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/MigrateTableSparkAction.java [hunk 16]
```diff
@@ -59,6 +60,7 @@ public class MigrateTableSparkAction extends BaseTableCreationSparkAction<Migrat
 
   private Identifier backupIdent;
   private boolean dropBackup = false;
+  private ExecutorService executorService;
```
Comment (nastra) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/MigrateTableSparkAction.java:63]: I believe we also need to handle shutdown of the executor in migrate + snapshot

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java [hunk 17]
```diff
@@ -102,6 +103,12 @@ public InternalRow[] call(InternalRow args) {
       action.tableLocation(snapshotLocation);
     }
 
+    if (!args.isNullAt(4)) {
+      int parallelism = args.getInt(4);
+      Preconditions.checkArgument(parallelism > 0, "Parallelism should be larger than 0");
+      action = action.executeSnapshotWith(executorService(parallelism, "table-snapshot"));
```
Comment (nastra) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/SnapshotTableProcedure.java:109]: shutdown of the executor should have been handled in `executorService`, but we need to do the same for the executor that is used for migrate. See also my other comment in https://github.com/apache/iceberg/pull/10037/files#r1654342618

File: data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java (hunk 15 above)
Comment (manuzhang) [reply data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java:219]: `ThreadPools.newWorkerPool` also returns an exiting executor service.

**Review by nastra — APPROVED** [review]
LGTM, thanks for getting this done @manuzhang

**Comment by puchengy** [comment]
Leaving an idea for a further speed up for table w/ data skewness on partition level: we can further divide files from a given partition into X number of buckets.

**Comment by manuzhang** [comment]
@puchengy Feel free to open a new issue to track.
